exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/StatisticsAggrFunctions.java

```diff
@@ -0,0 +1,5516 @@
+   * log2m the better the accuracy.
+   * accuracy = 1.04/sqrt(2^log2m)
+   * where
+   * log2m - the number of bits to use as the basis for the HLL instance
+   */
+  @FunctionTemplate(name = "hll", scope = FunctionTemplate.FunctionScope.POINT_AGGREGATE)
+  public static class HllFieldReader implements DrillAggFunc {
+    @Param FieldReader in;
+    @Workspace ObjectHolder work;
+    @Output NullableVarBinaryHolder out;
+    @Inject OptionManager options;
+    @Inject DrillBuf buffer;
+    @Workspace IntHolder hllAccuracy;
+
+    @Override
+    public void setup() {
+      work = new ObjectHolder();
+      hllAccuracy.value = (int) options.getLong(org.apache.drill.exec.ExecConstants.HLL_ACCURACY);
+      work.obj = new com.clearspring.analytics.stream.cardinality.HyperLogLog(hllAccuracy.value);
+    }
+
+    @Override
+    public void add() {
+      if (work.obj != null) {
+        com.clearspring.analytics.stream.cardinality.HyperLogLog hll =
+            (com.clearspring.analytics.stream.cardinality.HyperLogLog) work.obj;
+        int mode = in.getType().getMode().getNumber();
+        int type = in.getType().getMinorType().getNumber();
+
+        switch (mode) {
+          case org.apache.drill.common.types.TypeProtos.DataMode.OPTIONAL_VALUE:
+            if (!in.isSet()) {
+              hll.offer(null);
+              break;
+            }
+            // fall through //
+          case org.apache.drill.common.types.TypeProtos.DataMode.REQUIRED_VALUE:
+            switch (type) {
+              case org.apache.drill.common.types.TypeProtos.MinorType.VARCHAR_VALUE:
+                hll.offer(in.readText().toString());
+                break;
+              case org.apache.drill.common.types.TypeProtos.MinorType.BIGINT_VALUE:
+                hll.offer(in.readLong());
+                break;
+              default:
+                work.obj = null;
+            }
+            break;
+          default:
+            work.obj = null;
+        }
+      }
+    }
+
+    @Override
+    public void output() {
+      if (work.obj != null) {
+        com.clearspring.analytics.stream.cardinality.HyperLogLog hll =
+            (com.clearspring.analytics.stream.cardinality.HyperLogLog) work.obj;
+
+        try {
+          byte[] ba = hll.getBytes();
+          out.buffer = buffer.reallocIfNeeded(ba.length);
+          out.start = 0;
+          out.end = ba.length;
+          out.buffer.setBytes(0, ba);
+          out.isSet = 1;
+        } catch (java.io.IOException e) {
+          throw new org.apache.drill.common.exceptions.DrillRuntimeException("Failed to get HyperLogLog output", e);
+        }
+      } else {
+        out.isSet = 0;
+      }
+    }
+
+    @Override
+    public void reset() {
+      work.obj = new com.clearspring.analytics.stream.cardinality.HyperLogLog(hllAccuracy.value);
+    }
+  }
+
+  @FunctionTemplate(name = "ndv", scope = FunctionTemplate.FunctionScope.POINT_AGGREGATE)
```

Comment:
There is the "ndv" functions and the "hll" functions but only the latter are actually going to be used for the ANALYZE command, so let's only keep that.   
A related comment is that since we are computing the approximate count distinct, in the future we may externalize it using the SQL standard function **approx_count_distinct** that Oracle, MS SQL Server and others support.  So, we anyways should remove "ndv" since it can be misinterpreted to give actual num distinct values rather than approximate.

Reply:
We are using NDV functions so I changed that to approx_count_distinct (the merge is done in the NDVMergedStatistic). The HLL functions are currently unused but since we plan on adding support for incremental/partition stats it would be good to keep them especially since they were tested. Please let me know your thoughts.

Reply:
I am not sure I understood 'HLL functions are currently unused...' .  In Phase 1 aren't you using the HLL function and sending the NullableVarBinary to the Phase 2 where it is merged with others ?  Only in Phase 2 the NDV is being computed as a single numeric value.

Reply:
Yes, you are right. I forgot that we agreed to only use parallel stats so NDV will not be used. By HLL functions, I meant HLL_MERGE & HLL_DECODE functions. The HLL functions are used in the first phase.

Reply:
Right, so in my original comment I was likely referring to the NDV functions in the first phase.  Why    keep those unused functions.

Reply:
Removed the unused NDV functions